README.md
<h1 align="center">
 <a href="https://www.benbox.org/">
 <picture>
  <source media="(prefers-color-scheme: dark)" srcset="images/Logo.png">
  <img height="70" src="images/Logo.png">
 </picture>
 </a>
 <br>
Car Pool
</h1>
<p align="center">
find, propose and book business trips to share resources
<p align="center">
  <a href="https://www.benbox.org/Car_Pool/Introduction.html"><img src="https://img.shields.io/badge/View%20Documentation-Docs-yellow"></a>
  <a href="https://github.com/DrBenjamin/Car_Pool"><img src="https://img.shields.io/github/stars/DrBenjamin/Car_Pool" /></a>
</p>

# Car Pool App

The Car Pool App gives the staff the opportunity to find, propose and book business trips to share resources such as a vehicle, fuel and drivers to reduce costs, eco-system impact and to increase the opportunities to reach a destination.

## Installation and configuration of all needed Software

All Software which is used to run the **Car Pool App** is **Open Source**. Please be aware of different licenses with varying policies.

### Installation of Python and Streamlit

Install **[Streamlit &amp; Python](https://docs.streamlit.io/library/get-started/installation)** to run the source code locally. A virtual Python environment like **[Anaconda](https://anaconda.org/conda-forge/download)** / **[Miniconda](https://docs.conda.io/en/latest/miniconda.html)** is highly recommend.

#### Getting the source code and install dependencies

Clone the *repository* of **Car Pool** with following command:

```bash
git clone https://github.com/DrBenjamin/Car_Pool.git
```

After that you need to install some *Python libraries*. To do so use the `requirements.txt` file with:

```bash
cd Car_Pool
python -m pip install --upgrade -r requirements.txt
```

### Configuration of Streamlit config files

First make a directory `.streamlit`. After that create the file `.streamlit/config.toml`. Here you define the *theming* and some *Streamlit server behaviour* flags:

#### Theming and Streamlit server behaviour

Theming can be customized with these configuration:

```toml
# Theming
[theme]
primaryColor = "#F63366"
backgroundColor = "#FFFFFF"
secondaryBackgroundColor = "#F0F2F6"
textColor = "#262730"
font = "sans serif"

# Disable stats collecting
[browser]
gatherUsageStats = false

# Streamlit server behaviour
[server]
headless = true
```

#### Activating Secure Socket Layer (ssl)

If you want a secure connection (https), you need to generate the **[OpenSSL](https://slproweb.com/products/Win32OpenSSL.html)** certificate and the public key:

```bash
openssl genrsa 2048 > host.key
chmod 400 host.key
openssl req -new -x509 -nodes -sha256 -days 365 -key host.key -out host.cert
```

Streamlit brings secure connection out of the box, you just need to add these two lines to `.streamlit/config.toml`:

```toml
# Server certificate file for connecting via HTTPS. Must be set at the same time as "server.sslKeyFile"
sslCertFile = "<path-to-file>/host.cert"

# Cryptographic key file for connecting via HTTPS. Must be set at the same time as "server.sslCertFile"
sslKeyFile = "<path-to-file>/host.key"
```

#### Adding Mapbox API key

```toml
# Custom Mapbox token for elements like st.pydeck_chart and st.map
[mapbox]
token = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

### Hidden configuration file

Now create the file `.streamlit/secrets.toml` where you define some customisations and the *user / password* combinations:

```toml
# Customization
[custom]
organisation = "Company Name"
organisation_abbreviation = "CN"
logo_image = "images/Logo_Header.png"
sidebar_image = "images/Logo.png"

# User management
[passwords]
# Follow the rule: username = "password"
car = "pool"
```

### MySQL Server

A local **MySQL Server** is needed to run **HR Staff Portal**. Please install **MySQL Community Server** on your system (**[Windows](https://dev.mysql.com/downloads/mysql/)**, **[Ubuntu Linux](https://dev.mysql.com/doc/refman/8.0/en/binary-installation.html)**) or on Raspberry Pi use **MariaDB**:

```bash
sudo apt-get install mariadb-server mariadb-client
```

Use **[MySQL Workbench](https://dev.mysql.com/downloads/workbench/)** to configure the databases and user rights. Create a Schema with the name `carpool` and import the SQL sample file.

#### User, Schema and table structures

Name the database schema `carpool`. Create a user with access to it. Use the definded password in the ``.streamlit/secrets.toml`` file. The table structures and sample data are defined in the `files/carpool_dump.sql` file.

Table structures:

```sql
CREATE TABLE `cities` (
  `ID` int NOT NULL,
  `CITY` varchar(99) DEFAULT NULL,
  `LAT` varchar(45) DEFAULT NULL,
  `LON` varchar(45) DEFAULT NULL,
  PRIMARY KEY (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `distances` (
  `ID` int NOT NULL,
  `CITY` varchar(99) DEFAULT NULL,
  `City name 1` int DEFAULT NULL,
  `City name 2` int DEFAULT NULL,
  ...
  PRIMARY KEY (`ID`),
  UNIQUE KEY `ID_UNIQUE` (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `gamification` (
  `ID` int NOT NULL,
  `NAME` varchar(99) DEFAULT NULL,
  `POINTS` int DEFAULT NULL,
  PRIMARY KEY (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `routes` (
  `ID` int NOT NULL,
  `CITY` varchar(99) DEFAULT NULL,
  `City name 1` varchar(200) DEFAULT NULL,
  `City name 2` varchar(200) DEFAULT NULL,
  ...
  PRIMARY KEY (`ID`),
  UNIQUE KEY `ID_UNIQUE` (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `trips` (
  `ID` int NOT NULL,
  `DRIVER` varchar(99) DEFAULT NULL,
  `PHONE` varchar(99) DEFAULT NULL,
  `MAIL` varchar(99) DEFAULT NULL,
  `DEPARTURE` varchar(99) DEFAULT NULL,
  `DESTINATION` varchar(99) DEFAULT NULL,
  `DATE` datetime DEFAULT NULL,
  `START` datetime DEFAULT NULL,
  `ARRIVAL` datetime DEFAULT NULL,
  `SEATS` int DEFAULT '1',
  `REQUEST` tinyint DEFAULT '0',
  `FEMALE` tinyint DEFAULT '0',
  `FEMALE_GUESTS` tinyint DEFAULT '0',
  PRIMARY KEY (`ID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

Use the `files/Car_Pool_Sample.xlsx` Excel document to prepare location and routes data for your country. Export the sheets as csv and import to your database tables accordingly.

#### MySQL server configuration in Streamlit

In the `.streamlit/secrets.toml` you define the MySQL server settings:

```toml
# MySQL configuration
[connections.sql]
type = "sql"
dialect = "mysql"
host = "127.0.0.1"
port = 3306
database = "carpool"
username = "car"
password = "pool"
```

## Testing

To run an automated test just run the following command:

```bash
python files/Testing.py
```

## Build Windows help

To rebuild the Windows help file (after changes in the App and documentation 'Markdown' files), you first have to run the following command:

```bash
python files/Md2Html.py
```

to create the updated html files. Replace the images. Now rebuild the Help with **[HTML Help Workshop](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/htmlhelp/microsoft-html-help-downloads)**. Open the `docs/Html/Car_Pool.hhp` project file with the **HTML Help Workshop** Application and compile the `Car_Pool.chm` document. The help is accessible in the App through pressing the `F1` button.

## Run Application

To run the application just run the following command:

```bash
python -m streamlit run Car_Pool.py
```

### Run Application within Docker container

You can also run the Application within docker container without the need to install dependencies on the local system.

#### Install Docker

First install Docker on your machine with the following command:

```bash
# Install Docker
curl -fsSL https://get.docker.com -o get-docker.sh
sudo sh get-docker.sh

# Give super-user rights
sudo usermod -aG docker pi
```

#### Build and run Docker container

Now you can run the following commands to run the Streamlit app and the pre-configured MariaDB SQL server as Docker containers:

```bash
# Create shared network
docker network create carpool-network

# Build and run MariaDB SQL server
docker-compose build --no-cache && docker-compose up -d
# Connect database to Streamlit App
docker network connect carpool-network carpool-db-1

# Build Streamlit app
docker build --no-cache -t streamlit .
# Run Streamlit app
docker run --name carpool -p 8501:8501 --network=carpool-network streamlit

# Check network support of the 2 docker container
docker network inspect carpool-network # If the database connection fails, 
# please check the IP address of the database container and put it in `secrets.toml`
# under [mysql] host = "xxx.xxx.xxx.xxx" and re-build the Streamlit docker container!
```

## Contact

If you have any questions, please reach out to me at:

[ben@benbox.org](mailto:ben@benbox.org)

## Demo Application

 ```toml
user: car
password: pool
 ```

[![Open in Streamlit][share_badge]][share_link]

[share_badge]: https://static.streamlit.io/badges/streamlit_badge_black_white.svg
[share_link]: https://carpool.streamlit.app/


CHANGELOG.md
# Changelog

## [1.0.0](https://github.com/SeriousBenEntertainment/Car_Pool) - 2023-11-01

``` Version 1.0.0 Release```

### Functionality

* Car Pool
    * Enter a trip [Driving]
	* Search for a trip [Hitchhiking]
	* Book a trip  [Booking]
	* Gamification [Highscore]
    * Help (F1)


files/Md2Html.py
##### `files/Md2Html.py`
##### Car Pool
##### Open-Source, hosted on https://github.com/SeriousBenEntertainment/Car_Pool
##### Please reach out to ben@benbox.org for any questions
#### Loading needed Python libraries
import markdown
import os
import re

# Get all .md files in the `docs` folder
md_files = [f for f in os.listdir('docs') if f.endswith('.md')]
print('Markdown documents found: ' + ', '.join(md_files))

# Convert each .md file to .html
for md_file in md_files:
    with open("docs/" + md_file, "r", encoding = "utf8") as file:
        md_source = file.read()
    md_source = markdown.markdown(re.sub(r'[^\x00-\x7F]+|\x0c',' ', md_source))
    html_file = md_file.replace('.md', '.html')
    with open("docs/Html/" + html_file, "w") as file:
        file.write(md_source)
        print(html_file, ' updated!')

# Convert README, LICENSE and CHANGELOG documents
md_files = ['CHANGELOG.md', 'LICENSE']
for md_file in md_files:
    with open(md_file, "r", encoding = "utf8") as file:
        md_source = file.read()
        # Remove first 17 lines of `README.md`
        if md_file == "README.md":
            md_source = '\n'.join(md_source.split('\n')[17:])

    md_source = markdown.markdown(re.sub(r'[^\x00-\x7F]+|\x0c',' ', md_source))
    if md_file == "LICENSE":
        html_file = md_file + '.html'
    else:
        html_file = md_file.replace('.md', '.html')
    with open("docs/Html/" + html_file, "w") as file:
        file.write(md_source)
        print(html_file, ' updated!')

# Copy all png files to `docs/Html` folder
png_files = [f for f in os.listdir('docs') if f.endswith('.png')]
for png_file in png_files:
    with open("docs/" + png_file, "rb") as file:
        png_source = file.read()
    with open("docs/Html/" + png_file, "wb") as file:
        file.write(png_source)
        print(png_file, ' updated!')


files/Testing.py
##### `files/Testing.py`
##### Car Pool
##### Open-Source, hosted on https://github.com/SeriousBenEntertainment/Car_Pool
##### Please reach out to ben@benbox.org for any questions
#### Loading needed Python libraries
import streamlit as st
from streamlit.testing.v1 import AppTest
from datetime import date
from datetime import timedelta
from datetime import time

at = AppTest.from_file("Car_Pool.py")
at.run()
assert not at.exception

# Start testing
at.sidebar.text_input("username").input(list(st.secrets["passwords"].keys())[0]).run()
at.sidebar.text_input("password").input(list(st.secrets["passwords"].values())[0]).run()

# Tab `Driving``
at.text_input("name_driving").input("Benjamin Gross").run()
at.selectbox("sex_driving").select("male").run()
at.text_input("phone_driving").input("+265 98 347 6744").run()
at.text_input("mail_driving").input("benjamin.gross@giz.de").run()
at.selectbox("dep_driving").select("Dedza").run()
at.selectbox("des_driving").select("Balaka").run()
at.date_input("date_driving").set_value(date.today()).run()
at.time_input("stime_driving").set_value(time(12, 0)).run()
at.time_input("etime_driving").set_value(time(14, 0)).run()
at.number_input("num_driving").set_value(3).run()
at.selectbox("guest_driving").select("all sexes").run()
at.button[1].click().run() # Write to database

# Tab `Hitchhiking``
at.date_input("rstart").set_value(date.today()).run()
at.date_input("rend").set_value(date.today() + timedelta(days = 60)).run()
at.selectbox("dep_hitchhiking").select("Dedza").run()
at.selectbox("des_hitchhiking").select("Blantyre").run()
at.checkbox("fem_hitchhiking").check().run()
at.checkbox("fem_g_hitchhiking").check().run()

# Tab `Booking`
at.text_input("name_booking").input("Benjamin Gross").run()
at.number_input("num_booking").set_value(1).run()
at.button("book1").click().run() # Write to database

# Open help
at.button[0].click().run()

Car_Pool.py
##### `Car_Pool.py`
##### Car Pool
##### Open-Source, hosted on https://github.com/DrBenjamin/Car_Pool
##### Please reach out to ben@benbox.org for any questions
#### Loading needed Python libraries
import streamlit as st
import pandas as pd
import numpy as np
from datetime import datetime
from datetime import date
from datetime import time
from datetime import timedelta
# Mac SQL support
import pymysql
pymysql.install_as_MySQLdb()
# Prevents MySQL error
from sqlalchemy import text
import sys
sys.path.insert(1, "files/")
from functions import header
from functions import check_password
from functions import landing_page


## Streamlit initial setup
try:
    desc_file = open('DESCRIPTION', 'r')
    lines = desc_file.readlines()
    print(lines[3])
    st.set_page_config(
        page_title = "Car Pool",
        page_icon = st.secrets["custom"]["sidebar_image"],
        layout = "centered",
        initial_sidebar_state = "expanded",
        menu_items = {
            'Get Help': 'https://www.benbox.org',
            'Report a bug': 'https://github.com/DrBenjamin/Car_Pool/issues',
            'About': '**Car Pool** (' + lines[3] + ')\n\n The Car Pooling App gives the ' + st.secrets["custom"]["organisation_abbreviation"] + ' staff the opportunity to find, propose and book business trips to share resources such as a vehicle, fuel and drivers to reduce costs, eco-system impact and to increase the opportunities to reach a destination.'
        }
    )
except Exception as e:
    print(e)


## Initialization of session states
if ('logout' not in st.session_state):
    st.session_state['logout'] = False
if ('header' not in st.session_state):
    st.session_state['header'] = True
if ('index' not in st.session_state):
    st.session_state['index'] = 1
if ('booking' not in st.session_state):
    st.session_state['booking'] = []


## Functions
# Function: lastID = checks for last ID number in Table (to add data after)
def lastID():
    try:
        id = connection.query("SELECT MAX(ID) FROM carpool.trips;")
        id = int(id['MAX(ID)'].values[0]) + 1
    except Exception as e:
        id = 1
        print('No ID found ', e)
    return id

# Function: lastID = checks for last ID number in Table (to add data after)
def lastPOINTS():
    try:
        id = connection.query("SELECT MAX(ID) FROM carpool.gamification;")
        id = int(id['MAX(ID)'].values[0]) + 1
    except Exception as e:
        id = 1
        print('No ID found ', e)
    return id


## Logged in state
if check_password():
    # Header
    header(title = 'Car Pool', data_desc = 'car pooling', expanded = st.session_state['header'])
    st.title('Car Pool')
    st.write('The Car Pool App gives ' + st.secrets['custom']['organisation'] + ' staff the opportunity to find, propose and book business trips to share resources such as a vehicle, fuel and drivers to reduce costs, eco-system impact and to increase the opportunities to reach a destination. Now feel free to:')
    st.write('- enter a trip (🚗 **Driving**), if you can offer at least one seat;')
    st.write('- or search for a trip (👍🏿 **Hitchhiking**) for your purpose;')
    st.write('- or book a matching trip (📅 **Booking**) you found.')
    
    # Open databank connection (Streamlit)
    connection = st.connection(name = 'sql', type = 'sql', autocommit = True)

    # Get trips data, (NO CACHING!)
    databank = connection.query("SELECT ID, DRIVER, PHONE, MAIL, DEPARTURE, DESTINATION, DATE, START, ARRIVAL, SEATS, REQUEST, FEMALE, FEMALE_GUESTS FROM `carpool`.`trips`;", ttl = 0)
    databank = databank.set_index('ID')

    # Get cities data (cached)
    databank_cities = connection.query("SELECT ID, CITY, LAT, LON FROM `carpool`.`cities`;")
    databank_cities = databank_cities.set_index('ID')

    # Get distances data (cached)
    databank_distances = connection.query("SELECT * FROM `carpool`.`distances`;")
    databank_distances = databank_distances.set_index('ID')

    # Get routes data (cached)
    databank_routes = connection.query("SELECT * FROM `carpool`.`routes`;")
    databank_routes = databank_routes.set_index('ID')

    # Get points (gamification)
    databank_points = connection.query("SELECT ID, NAME, POINTS FROM `carpool`.`gamification`;")
    databank_points = databank_points.set_index('ID')
    st.sidebar.subheader('Highscore 🏆')
    databank_points = databank_points.sort_values('POINTS', ascending = False)
    st.sidebar.dataframe(databank_points[['NAME', 'POINTS']], column_config = {
        "NAME": st.column_config.TextColumn("Name"),
        "POINTS": st.column_config.NumberColumn("Points")}, hide_index = True)
    
    # Menu
    font_css = """<style>button[data-baseweb = "tab"] > div[data-testid="stMarkdownContainer"] > p {font-size: 18px;}</style>"""
    st.write(font_css, unsafe_allow_html = True)
    tab1, tab2, tab3 = st.tabs(["🚗 **Driving** :gray[(*enter a trip*)]", "👍🏿 **Hitchhiking** :gray[(*see future trips*)]", "📅 **Booking** :gray[(*book a trip*)]"])

    # tab `Driving`
    with tab1:
        with st.form("Driving", clear_on_submit = True):
            st.write("")
            st.title('Driving')
            st.subheader('Enter a trip')
            st.write('Please enter all data precisely, for a trip, you can offer at least one seat in your car.')
            name = st.text_input('Name', key = "name_driving")
            sex = st.selectbox('Sex', options = ['female', 'male'], index = 0, key = "sex_driving")
            if sex == 'female':
                female = 1
            else:
                female = 0
            phone = st.text_input('Phone', key = "phone_driving")
            mail = st.text_input('Mail', key = "mail_driving")
            dep_list = np.array([])
            dep_list = np.insert(databank_cities['CITY'].values, 0, 'All departures')
            dep = st.selectbox('Departure', options = dep_list, index = 1, key = "dep_driving")
            des_list = np.array([])
            des_list = np.insert(databank_cities['CITY'].values, 0, 'All destinations')
            des = st.selectbox('Destination', options = des_list, index = 3, key = "des_driving")
            date_t = st.date_input('Date', key = "date_driving")
            # Add time to date
            date_t = datetime.combine(date_t, time(0, 0)) # Convert date to datetime
            time_start = st.time_input('Start time', value = time(11, 30), key = "stime_driving")
            time_start = datetime.combine(date_t, time_start) # Convert time to datetime
            time_end = st.time_input('Approx. arrival time', value = time(12, 45), key = "etime_driving")
            time_end = datetime.combine(date_t, time_end) # Convert time to datetime
            seats = st.number_input('Seats', min_value = 1, max_value = 6, value = 1, key = "num_driving")
            guests = st.selectbox('Guests', options = ['all sexes', 'only female!'], index = 0, key = "guest_driving")
            if female == 1:
                if guests == 'only female!':
                    female_guests = 1
                else:
                    female_guests = 0
            else:
                female_guests = 0

            # Submit button
            submitted = st.form_submit_button('Submit', help = "Write your trip to the database.")
            if submitted:
                try:
                    # Writing to database
                    with connection.session as session:
                        id = lastID()
                        session.execute(text("INSERT INTO `carpool`.`trips`(ID, DRIVER, PHONE, MAIL, DEPARTURE, DESTINATION, DATE, START, ARRIVAL, SEATS, REQUEST, FEMALE, FEMALE_GUESTS) VALUES (:id, :name, :phone, :mail, :dep, :des, :date_t, :time_start, :time_end, :seats, :request, :female, :female_guests);"), {"id": id, "name": name, "phone": phone, "mail": mail, "dep": dep, "des": des, "date_t": date_t, "time_start": time_start, "time_end": time_end, "seats": seats, "request": 0, "female": female, "female_guests": female_guests})
                        session.commit()
                        print('Updated database, table `trips`')
                        st.toast('Updated database!', icon = '🎉')
                    # Updating `gamification` table
                    with connection.session as session:
                        id = lastPOINTS()
                        if name in databank_points['NAME'].values:
                            session.execute(text("UPDATE `carpool`.`gamification` SET POINTS = :points WHERE NAME = :name;"), {"points":  int(databank_points.loc[databank_points['NAME'] == name]['POINTS'].values[0]) + 10, "name": name})
                            session.commit()
                        else:
                            session.execute(text("INSERT INTO `carpool`.`gamification`(ID, NAME, POINTS) VALUES (:id, :name, :points);"), {"id": id, "name": name, "points": 10})
                            session.commit()
                        print('Updated database, table `gamification`')
                except Exception as e:
                    print('No Update of database', e)

    # tab `Hitchhiking`
    with tab2:
        with st.expander('', expanded = True):
            st.title('Hitchhiking')
            st.subheader('Look for trips')
            st.write('Choose the range of dates in between you want to travel to find a matching trip (*booking will be done by these results*).')

            # Set range of dates
            range_date = []
            date_start = st.date_input('Range start', key = "rstart")
            range_date.append(date_start)
            date_end = st.date_input('Range end', value = date.today() + timedelta(days = 60), key = "rend")
            range_date.append(date_end)
            st.info('Choose the same date twice if you are looking just for a single day.', icon = "ℹ️")
            
            # Set the Departure and Destination
            st.write('Choose the departure and destination of your travel.')

            # Add option all at the start of the list
            dep_list = np.array([])
            dep_list = np.insert(databank_cities['CITY'].values, 0, 'All departures')
            dep = st.selectbox('Departure', options = dep_list, index = 1, key = "dep_hitchhiking")
            des_list = np.array([])
            des_list = np.insert(databank_cities['CITY'].values, 0, 'All destinations')
            des = st.selectbox('Destination', options = des_list, index = 3, key = "des_hitchhiking")

            # Distance calculation
            if dep != 'All departures' and des != 'All destinations':
                distance = databank_distances.loc[databank_distances['CITY'] == dep][des].values[0]
                st.write('Distance of this trip is about ', str(distance), 'km.')

            # Female only trips
            fem = st.checkbox('Female driver?', key = "fem_hitchhiking")
            if fem:
                fem_g = st.checkbox('Only female guests?', key = "fem_g_hitchhiking")
            else:
                fem_g = False


            ## Search for matching trips in the future
            actual_data = []
            # Get directions on the route in the right order
            try:
                direction = databank_routes.loc[databank_routes['CITY'] == dep][des].values[0].replace(" ", "").split(',')
            except:
                direction = []
                direction.append("")
            # Add `Departure` to direction
            if direction[0] == "":
                direction[0] = dep
            else:
                direction.insert(0, dep)
            # Add `Destination` to direction 
            direction.append(des)
            index_dep = direction.index(dep)
            index_des = direction.index(des)
            direction_direction = index_dep - index_des
            # Find the matching trips
            for idx, row in databank.iterrows():
                # Get cities on the route in the right order
                cities = databank_routes.loc[databank_routes['CITY'] == row['DEPARTURE']][row['DESTINATION']].values[0].replace(" ", "").split(',')
                # Add `Departure` to cities
                if cities[0] == "":
                    cities[0] = row['DEPARTURE']
                else:
                    cities.insert(0, row['DEPARTURE'])
                # Add `Destination` to cities
                cities.append(row['DESTINATION'])
                # Calculate order
                try:
                    index1 = cities.index(dep)
                    index2 = cities.index(des)
                    direction_cities = index1 - index2
                except:
                    direction_cities = 0
                # Add trips to the list if matching
                if (datetime.date(row['DATE']) >= range_date[0] and datetime.date(row['DATE']) <= range_date[1]) and int(row['SEATS'] - row['REQUEST']) >= 1:
                    if dep == 'All departures' or dep in cities:
                        if des == 'All destinations' or des in cities:
                            # Check if the array `cities` is in the same direction of array `direction`
                            if (direction_cities < 0 and direction_direction < 0) or (direction_cities > 0 and direction_direction > 0) or des == 'All destinations':
                                if len(dep) > 0:    
                                    if fem:
                                        if row['FEMALE'] == 1:
                                            if fem_g:
                                                if row['FEMALE_GUESTS'] == 1:
                                                    actual_data.append(row)
                                            else:
                                                actual_data.append(row)
                                    else:
                                        actual_data.append(row)
            actual_data = pd.DataFrame(actual_data)
            
            # Show data
            try:
                st.subheader('List view of trips')
                st.write('For your selection, there are ' + str(len(actual_data.index)) + ' trips in the database.')
                try:
                    if len(actual_data) > 0:
                        if len(actual_data) == 1:
                            st.info('Go to **Booking** page to book this trip!', icon = "‼")
                        else:
                            st.info('Go to **Booking** page to book one of these trips!', icon = "‼")
                except:
                    print('No data.')
                st.dataframe(actual_data[['DEPARTURE', 'DESTINATION', 'DATE', 'START', 'ARRIVAL', 'SEATS', 'REQUEST']], column_config = {
                    "DEPARTURE": st.column_config.TextColumn("From (Departure)"),
                    "DESTINATION": st.column_config.TextColumn("To (Destination)"),
                    "DATE": st.column_config.DatetimeColumn("Date", format = "D MMM YYYY"),
                    "START": st.column_config.DatetimeColumn("Departure", format = "HH:mm"),
                    "ARRIVAL": st.column_config.DatetimeColumn("Arrival", format = "HH:mm"),
                    "SEATS": st.column_config.NumberColumn("Seats"),
                    "REQUEST": st.column_config.NumberColumn("Already requested")}, hide_index = True)
                i = 0
                for trip in actual_data[['FEMALE', 'FEMALE_GUESTS']].values:
                    i += 1
                    if not fem:
                        if trip[0] == 1:
                            st.info('Trip ' + str(i) + ' is offered by a female driver.', icon = "🧍🏾‍♀️")
                    if not fem_g:
                        if trip[1] == 1:
                            st.info('Trip ' + str(i) + ' is for female guests only!', icon = "👭")
                st.session_state['booking'] = actual_data
            # No data available  
            except:
                st.warning(body = 'No Trips in this range!', icon = "🚨")
                st.info('Choose other selections if possible.', icon = "ℹ️")

    # tab `Booking`
    with tab3:
        with st.expander('', expanded = True):
            st.title('Booking')
            st.subheader('Book a trip')
            st.write('Here you book a trip (*you need to contact the driver afterwards*).')
            name = ""
            name = st.text_input('Name (*needed to add your points to the Highscore* 🏆)', key = "name_booking")
            try:
                st.dataframe(st.session_state['booking'][['DEPARTURE', 'DESTINATION', 'DATE', 'START', 'ARRIVAL', 'SEATS', 'REQUEST']], column_config = {
                        "DEPARTURE": st.column_config.TextColumn("From (Departure)"),
                        "DESTINATION": st.column_config.TextColumn("To (Destination)"),
                        "DATE": st.column_config.DatetimeColumn("Date", format = "D MMM YYYY"),
                        "START": st.column_config.DatetimeColumn("Departure", format = "HH:mm"),
                        "ARRIVAL": st.column_config.DatetimeColumn("Arrival", format = "HH:mm"),
                        "SEATS": st.column_config.NumberColumn("Seats"),
                        "REQUEST": st.column_config.NumberColumn("Already requested")}, hide_index = True)
                # Book trip
                seats = st.number_input('Seats', min_value = 1, max_value = 6, value = 1, key = "num_booking")
                i = 0
                for idx, row in st.session_state['booking'].iterrows():
                    if seats <= int(row["SEATS"] - row["REQUEST"]):
                        i += 1
                        if st.button('Book trip ' + str(i) + ' (' + str(row["DATE"].strftime('%d-%m-%Y' + ')')), key = "book" + str(i), help = 'Book the trip '):
                            # Update database
                            try:
                                with connection.session as session:
                                    session.execute(text("UPDATE `carpool`.`trips` SET REQUEST = :seats WHERE ID = :id;"), {"seats": int(seats + row["REQUEST"]), "id": idx})
                                    session.commit()
                                    print('Updated database, table `trips`')
                                    st.toast("Please contact the driver!", icon = '💬')
                                    st.session_state['booking'] = []
                                if name != "":
                                    with connection.session as session:
                                        id = lastPOINTS()
                                        if name in databank_points['NAME'].values:
                                            session.execute(text("UPDATE `carpool`.`gamification` SET POINTS = :points WHERE NAME = :name;"), {"points":  int(databank_points.loc[databank_points['NAME'] == name]['POINTS'].values[0]) + 5, "name": name})
                                            session.commit()
                                        else:
                                            session.execute(text("INSERT INTO `carpool`.`gamification`(ID, NAME, POINTS) VALUES (:id, :name, :points);"), {"id": id, "name": name, "points": 5})
                                            session.commit()
                                        print('Updated database, table `gamification`')
                                st.write('**Driver:** ' + row['DRIVER'] + ' (' + row['PHONE'] + ')')
                                st.write('**Mail:** ' + row['MAIL'])
                                st.info('Please contact the driver!', icon = "💬")
                            except Exception as e:
                                print('No request sent', e)
                if i == 0:
                    st.info('Lower the amount of seats you are requesting.', icon = "ℹ️")       
            except:
                st.warning(body = 'No trips selected!', icon = "🚨")
                st.info('Go to `Hitchhiking` to find a fitting trip.', icon = "ℹ️")     
            st.subheader('Map view of trips')
            places = np.concatenate((st.session_state['booking']['DEPARTURE'].unique(), st.session_state['booking']['DESTINATION'].unique()), axis = 0)
            df = pd.DataFrame([], columns = ['lat', 'lon'])
            for place in places:
                for cordinate in databank_cities['CITY']:
                    if place == cordinate:
                        lat = str(databank_cities.loc[databank_cities['CITY'] == cordinate]['LAT'].values)[1:-1]
                        lon = str(databank_cities.loc[databank_cities['CITY'] == cordinate]['LON'].values)[1:-1]
                        df.loc[len(df.index)] = [float(lat.replace("'", "")), float(lon.replace("'", ""))]
            # Show map
            st.map(df, size = 20, use_container_width = False)


## Not logged in state (Landing page)
else:
    landing_page(' Car Pool')
